### PR TITLE
[Distributed Strategy] GLM-4.5 model accuracy alignment

### PR DESCRIPTION
**Pull request by zhanghonggeng**
### PR Category

Distributed Strategy

### PR Types

New features

### Description

GLM-4.5 模型精度对齐（PaddleFleet vs Megatron-LM），所有对齐逻辑通过 use_accuracy_compatible 开关控制，开关关闭时不影响原有代码行为。

核心改动：

moe_utils.py — 实现确定性 permute/unpermute（gather+sum fp32 PyLayer），消除 scatter_add 非确定性累加序差异

moe_router.py — gate weight dtype 改为 bf16（对齐 MG 默认行为）；router scoring 使用 re-sigmoid 计算 gates_ori；TopK 归一化使用 fp64 求和；aux_loss 计算对齐 MG 公式；take_along_axis 替换为 gather_nd（对齐反向梯度路径）

moe_layer.py — 三路 clone（ThreePathCloneAlignMG PyLayer）控制 bf16 backward 累加顺序；小 M expert padding（M<17 → 32）消除 cublasLt tiny-M kernel 差异；EP>1 时强制 alltoall dispatcher

fused_bias_swiglu.py — 对齐模式下使用 eager SwiGLU forward/backward（非 fused），避免 jit_fuser kernel 数值差异

tensor_parallel/layers.py — forward 使用 weight.T.contiguous() + paddle.matmul 对齐 GEMM layout；expert weight 的 wgrad 使用 fp32 直接累加到 main_grad

language_loss.py — loss 求和使用 fp64 精度；模拟 EP allreduce 的 bf16 累加序

model_parallel_config.py — 对齐模式下自动启用 deterministic_mode=True

rotary_pos_embedding.py — 移除已无用的 use_accuracy_compatible 分支（CPU/GPU 计算 inv_freq 数值等价）

### 是否引起精度变化

否

### CODE REVIEW

File: src/paddlefleet/models/common/language_loss/language_loss.py [hunk 1]
```diff
@@ -430,10 +432,26 @@ def _cast_loss_func(logits, labels):
...
+                    _flat = loss.cast(paddle.float32).reshape([-1]) * lossmask
+                    loss_sum = (
+                        _flat.cast(paddle.float64).sum().cast(paddle.float32)
+                    )
+                    _count = lossmask.sum()
+                    import paddle.distributed as _pdist
+
+                    _ep_size = _pdist.get_world_size(
+                        group=get_expert_model_parallel_group()
+                    )
```
Comment (risemeup1111) [inline src/paddlefleet/models/common/language_loss/language_loss.py]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

`USE_ACCURACY_COMPATIBLE=1` 会让共享的语言模型 loss 进入这个分支，但 `get_expert_model_parallel_group()` 默认要求 EP group 已初始化；`expert_model_parallel_size` 默认是 1，非 MoE 或 EP=1 配置下这里会直接 assert，导致 accuracy-compatible 开关一打开就无法跑 loss。EP 未启用时应当按 `ep_size=1` 处理。

```suggestion
                    _ep_group = get_expert_model_parallel_group(
                        check_initialized=False
                    )
                    _ep_size = (
                        _pdist.get_world_size(group=_ep_group)
                        if _ep_group is not None
                        else 1
                    )
```

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 2]
```diff
@@ -484,6 +481,25 @@ def _cal_seq_aux_loss(
...
+            ).indices.cast("int64")
+            _aux_routing_map_pf = paddle.zeros_like(
+                _probs_2d_pf
+            ).put_along_axis_(
+                _aux_top_idx_pf,
+                paddle.to_tensor(1.0, dtype=_probs_2d_pf.dtype),
+                axis=-1,
+            )
+            _aux_routing_map_pf = _aux_routing_map_pf.reshape(routing_map.shape)
+            routing_map = _aux_routing_map_pf
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_router.py]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这里在 accuracy-compatible 分支里把 `routing_map` 直接从 `probs` 重新 top-k 出来，但没有保留原始的行级 valid mask。对于 padding / invalid token，`probs` 是全 0，`topk` 仍会硬选出若干 expert，后面 `routing_map.sum(...)` 和 `cost_coeff` 会把这些 padding token 也算进去，seq_aux_loss 会偏掉。

建议在重建后把原始“这一行是否有效”的 mask 乘回去，只保留真实 token 的 routing：

```suggestion
            _aux_routing_map_pf = _aux_routing_map_pf.reshape(routing_map.shape)
            row_mask = (
                routing_map.cast("int64").sum(axis=-1, keepdim=True) > 0
            ).cast(_probs_2d_pf.dtype)
            routing_map = _aux_routing_map_pf * row_mask
```

**Comment by risemeup1111** [comment]
### Nyanpasu Review Board


| 序号 | 位置 | 优先级 | 状态 |
| - | - | - | - |
| 1 | - | - | ✅ |

<details>
<summary>点击展开 Review 规则</summary>
PR 评审规则：

- P0、P1 级别的评审必须提交新的 commit 进行修改；
- P2、P3 级别的评审可以通过评论进行修改。

状态说明：

| 已解决 | 无需修复 | 待修复 |
| - | - | - |
| ✅ | 🟡 | 🚧 |
</details>

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
我已把当前 head 上需要先修的两个问题写在行级评论里，细节都在 inline review comments 中。

当前还有一条构建作业受外部镜像拉取超时影响，这部分不算代码问题。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/fusions/fused_bias_swiglu.py [hunk 3]
```diff
@@ -21,6 +21,7 @@
 import paddle.nn.functional as F
 
 from paddlefleet.jit import jit_fuser
+from paddlefleet.transformer.moe.moe_utils import _use_accuracy_compatible
```
Comment (risemeup1111) [inline src/paddlefleet/fusions/fused_bias_swiglu.py]: | 状态 | 优先级 | 问题 | 位置 |
| --- | --- | --- | --- |
| 待处理 | P1 P1 | 顶层导入 `moe_utils` 会触发 `fused_bias_swiglu` 与 MoE 包初始化的循环导入。 | `src/paddlefleet/fusions/fused_bias_swiglu.py:24` |

File: src/paddlefleet/transformer/moe/moe_utils.py [hunk 4]
```diff
@@ -204,16 +332,50 @@ def permute(
...
+        expert_offsets[1:] = paddle.cumsum(tokens_per_expert, axis=0)
+        position_in_expert_T = rm_int.cumsum(axis=-1) - 1
+        global_position = (
+            position_in_expert_T + expert_offsets[:-1].unsqueeze(1)
+        ).T
+
+        rm_bool_token = routing_map.cast(paddle.bool)
+        per_token_k = rm_bool_token.cast("int64").sum(axis=-1)
+        topk_val = int(per_token_k.max().item())
+        if int(per_token_k.min().item()) != topk_val:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_utils.py]: | 状态 | 优先级 | 问题 | 位置 |
| --- | --- | --- | --- |
| 待处理 | P1 P1 | 固定 top-k 检查会让带 padding 的 accuracy-compatible MoE 输入直接抛错。 | `src/paddlefleet/transformer/moe/moe_utils.py:357` |

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
### Nyanpasu Review Board

| 状态 | 优先级 | 问题 | 位置 |
| --- | --- | --- | --- |
| 待处理 | P1 P1 | 顶层导入 `moe_utils` 会触发 `fused_bias_swiglu` 与 MoE 包初始化的循环导入。 | `src/paddlefleet/fusions/fused_bias_swiglu.py:24` |
| 待处理 | P1 P1 | 固定 top-k 检查会让带 padding 的 accuracy-compatible MoE 输入直接抛错。 | `src/paddlefleet/transformer/moe/moe_utils.py:357` |
| 待处理 | P3 P3 | PR 标题缺少仓库要求的 `[CLASS]` 前缀。 | 非行级：PR 标题 |
| 已解决 | P1 P1 | EP group 未初始化时 language loss 不再强制读取 EP group。 | `src/paddlefleet/models/common/language_loss/language_loss.py:443` |
| 已解决 | P1 P1 | `seq_aux_loss` 重建 routing map 时已保留原始有效行 mask。 | `src/paddlefleet/transformer/moe/moe_router.py:501` |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/model_parallel_config.py [hunk 5]
```diff
@@ -380,6 +381,9 @@ def __post_init__(self):
         if self.tensor_model_parallel_size <= 1:
             self.sequence_parallel = False
 
+        if os.environ.get("USE_ACCURACY_COMPATIBLE", "0") == "1":
+            self.deterministic_mode = True
```
Comment (risemeup1111) [inline src/paddlefleet/model_parallel_config.py:384]: P1 `USE_ACCURACY_COMPATIBLE=1` 这里只设置了 `deterministic_mode`，但当前对齐分支都读取默认仍为 `False` 的 `config.use_accuracy_compatible`，因此按 PR 描述只设置环境变量时 MoE/router/SwiGLU/loss 的 accuracy-compatible 分支不会生效；请在该 env 分支中同步打开该配置，或保留无循环依赖的运行时开关入口。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
本次复查确认循环导入问题已由后续 commit 修复；`moe_utils` 的 padding/top-k 问题仍未修复，并新增一处环境变量开关没有真正打开 accuracy-compatible 配置的问题。阻塞项仍需要通过新的 commit 处理。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/moe/moe_utils.py [hunk 6]
```diff
@@ -154,6 +154,97 @@ def _unpermute_fp32_accum(
...
+    global_position = position_in_expert_T + expert_offsets[:-1].unsqueeze(
+        1
+    )  # [num_experts, num_tokens]
+    global_position_per_token = global_position.T  # [num_tokens, num_experts]
+
+    topk = int(routing_map_bool.cast("int64").sum(axis=-1)[0].item())
+    valid_positions = global_position_per_token * routing_map_bool.cast("int64")
+    reverse_indices_flat = paddle.masked_select(
+        valid_positions, routing_map_bool
+    ).reshape([num_tokens * topk])
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_utils.py]: P1 padding 行仍会让 accuracy-compatible unpermute/permute 失效：这里从第 0 行取 `topk` 并按 `num_tokens * topk` reshape，而 permute backward 也用原始 `num_tokens * topk`；当前又允许 `k=0` padding 行，含 padding 的 `routing_map` 会 shape mismatch 或梯度错位，请改为按有效 token 数构造反向映射并把 padding 行输出/梯度补 0。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
复查后仍有一处 P1 需要通过新的 commit 修复；其余已复核的旧阻塞项已按当前代码和 PR 说明更新状态。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/moe/moe_utils.py [hunk 7]
```diff
@@ -154,6 +154,113 @@ def _unpermute_fp32_accum(
...
+    )  # [num_experts, num_tokens]
+    global_position = position_in_expert_T + expert_offsets[:-1].unsqueeze(
+        1
+    )  # [num_experts, num_tokens]
+    global_position_per_token = global_position.T  # [num_tokens, num_experts]
+
+    per_token_k = routing_map_bool.cast("int64").sum(axis=-1)
+    valid_token_indices = paddle.nonzero(per_token_k > 0).flatten()
+    valid_ks = per_token_k.index_select(axis=0, index=valid_token_indices)
+    topk = int(valid_ks[0].item())
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_utils.py:242]: P1 这里需要处理全 padding / 全 0 `routing_map`；当前 `valid_ks` 为空时会直接访问 `valid_ks[0]`。Router 会把 padding token 的 routing 行清零，AllToAll/combine_postprocess 仍会把这个 `routing_map` 传进 accuracy-compatible unpermute，因此全 padding microbatch 或 rank 会在这里崩溃。

File: src/paddlefleet/tensor_parallel/layers.py [hunk 8]
```diff
@@ -1049,7 +1056,24 @@ def backward(ctx, grad_output):
             else:
                 grad_weight = None
         else:
-            if wgrad_compute:
+            if (
```
Comment (risemeup1111) [inline src/paddlefleet/tensor_parallel/layers.py:1059]: P1 这个专家 `main_grad` 分支也需要受 `wgrad_compute` 保护。上面 `grad_output_buffer` 命中延迟 wgrad 时不会构造 `total_input`，但这里仍会进入累加并读取 `total_input`，开启 `defer_embedding_wgrad_compute` + `use_accuracy_compatible` 的专家线性层会在 backward 中报错并破坏延迟语义。

File: src/paddlefleet/fusions/fused_bias_swiglu.py [hunk 9]
```diff
@@ -376,14 +431,25 @@ def backward(ctx, grad_output):
...
+    ):
         input_for_backward = (
             input.to(paddle.float8_e4m3fn) if fp8_input_store else input
         )
         ctx.save_for_backward(input_for_backward, weights)
         ctx.ori_input_dtype = input.dtype
         ctx.fp8_input_store = fp8_input_store
         ctx.clamp_value = clamp_value
+        ctx.use_accuracy_compatible = use_accuracy_compatible
+        if use_accuracy_compatible:
```
Comment (risemeup1111) [inline src/paddlefleet/fusions/fused_bias_swiglu.py]: P1 `use_accuracy_compatible` 不应绕过 `clamp_value` 分支。MLP 的 `per_token_scale` 路径会同时传入 `activation_func_clamp_value` 和 `use_accuracy_compatible`，当前这里先返回 eager SwiGLU，导致已配置的 clamp 在 forward/backward 都失效。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
当前仍有几处会导致 accuracy-compatible 路径崩溃或跳过已有 clamp 语义的问题，需要先通过新的 commit 修复；细节见行级评论。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/moe/moe_utils.py (hunk 7 above)
Comment (zhanghonggeng) [reply src/paddlefleet/transformer/moe/moe_utils.py:242]: 已在 [moe_utils.py:236-240](https://webview-0934so43k0q60rq2foof7o5e1etqqmh3hlakaq56vamftlj63gaj.icoding.baidu-int.com/icoding/webview/comate/stable/1.108.2/PaddleFleet/src/paddlefleet/transformer/moe/moe_utils.py#L236-L240) 增加早返回：


per_token_k = routing_map_bool.cast("int64").sum(axis=-1)
valid_token_indices = paddle.nonzero(per_token_k > 0).flatten()
if valid_token_indices.numel() == 0:
    return paddle.zeros(restore_shape, dtype=permuted_tokens.dtype)
这样当整个 microbatch 或某个 rank 上所有 token 都是 padding 时：

不会访问空的 valid_ks[0]；
不会执行后续 min/max；
直接返回 [num_tokens, hidden] 的全零结果；
有效 token 的原有路径不变。

File: src/paddlefleet/tensor_parallel/layers.py (hunk 8 above)
Comment (zhanghonggeng) [reply src/paddlefleet/tensor_parallel/layers.py:1059]: 当前 [layers.py:1020-1032](https://webview-0934so43k0q60rq2foof7o5e1etqqmh3hlakaq56vamftlj63gaj.icoding.baidu-int.com/icoding/webview/comate/stable/1.108.2/PaddleFleet/src/paddlefleet/tensor_parallel/layers.py#L1020-L1032) 已经有：


elif ctx.gradient_accumulation_fusion:
    if wgrad_compute:
        ...
因此 grad_output_buffer 触发延迟 wgrad 时，wgrad_compute=False，不会进入 main_grad 累加分支，也不会读取没有构造的 total_input。

这条 review 对当前工作区代码已经满足，不需要额外修改。

Comment (risemeup1111) [reply src/paddlefleet/tensor_parallel/layers.py:1059]: 我同意你指出的 `elif ctx.gradient_accumulation_fusion` 分支已有 `wgrad_compute` 保护，但当前评论指向的是后面的非 fusion `else` 分支。`grad_output_buffer` 命中延迟时 `wgrad_compute=False`，上方不会构造 `total_input`，而这里的 accuracy-compatible expert `main_grad` 条件仍不包含 `wgrad_compute`；例如 `ColumnParallelLinear` 在 defer 开启时会传入 buffer，但调用这里时 `gradient_accumulation_fusion=False`。所以该分支仍需要受 `wgrad_compute` 保护。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
复查当前 head 后，`moe_utils.py` 的全 padding 早返回和 `WeightedSwiGLUFunction` 的 clamp 优先级已经修复；`layers.py` 的延迟 wgrad 路径仍需按原线程补上保护，PR 标题项也仍未处理。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — COMMENTED** [review]
复查当前 head 后，Nyanpasu 之前提出的 P1 行级问题都已通过新 commit 修复；当前只剩 PR 标题格式项需要处理，状态已更新到 Review Board。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — COMMENTED** [review]
复查当前状态后，Nyanpasu 之前提出的行级问题和 PR 标题格式都已解决；但 `Coverage Upload And Check` 仍失败，patch coverage 为 29.36%，低于 90% 阈值，因此本轮先不批准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — APPROVED** [review]
<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>
